src/nvim/api/ui.c

```diff
@@ -53,8 +53,18 @@ void remote_ui_disconnect(uint64_t channel_id)
 
 void nvim_ui_attach(uint64_t channel_id, Integer width, Integer height,
                     Dictionary options, Error *err)
-  FUNC_API_SINCE(1) FUNC_API_REMOTE_ONLY
+  FUNC_API_SINCE(1)
 {
+  if (is_internal_call(channel_id)) {
```

Comment:
This is a bit weird reuse of a function with a different purpose (the fact that you have to ignore the args demonstrates this). What would be wrong with `nvim_tui_start` and `nvim_tui_stop` ?

Reply:
You know what my answer is :) 

I strongly dislike special case functions in the API, I don't like sending people  to different entry points for related concepts.  

Ignoring the args is a compromise because API does not support optional args.  

It would be more weird to treat the TUI specially. That would be really unfortunate because it's an opportunity to emphasize the decoupled design and the fact that TUI is treated mostly like any other external UI.

Reply:
But my very point is that that this is a _weird_ and unfortunate special case for TUI. :)

This function never does ask an UI start. Rather it is _used by_ an intented UI to ask nvim to subscribe to UI events. The fact that args is ignored has noting to do with optional args, the point that it demonstrates an abstraction mismatch would stand even if the API had optional args. (because you can still pass the args, and they make no sense at all)

I'm not trying to be pedantic, as indeed the _actual_ meaning of this function has a straightforward generalization to internal use: by lua code to ask lua callback to be called on UI layer events (so that lua can draw externalized widgets on a private grid layer)

Also for rplugins it would be weird to not be able to call this function directly. Why should `vim.funcs.nvim_ui_detach` have a radically different meaning than `vim.api.ui_detach`. isn't `api.tui_stop` much more self-documenting and explicit?

Reply:
Good points.

Reply:
Good points. How do you feel about a `:tui` command, similar to the `:gui` command?

Reply:
In general I prefer new API functions, though
``` 
:tui /dev/pts/7
```
to take over a pty would have a nice ring to it :) Tough this probably shouldn't be builtin, rather a separate script could take over the pty, and then launch the future nvim tui client and connect to `NVIM_LISTEN_ADDRESS`.